In [1]:
from google.colab import userdata
userdata.get('HuggingFace')

'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'

In [2]:
!pip -q install -U transformers datasets accelerate evaluate scikit-learn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.7/10.7 MB 40.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.5/527.5 kB 16.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 32.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.6/47.6 MB 19.3 MB/s eta 0:00:00


In [3]:
import pandas as pd
import numpy as np

Príprava Slovak SA

In [4]:
#load SlovakSA datasets as dataframe
splits = {'train': 'train.csv', 'validation': 'dev.csv', 'test': 'test.csv'}

#save into separate DFs.
train_df = pd.read_csv("hf://datasets/DGurgurov/slovak_sa/train.csv")
val_df   = pd.read_csv("hf://datasets/DGurgurov/slovak_sa/dev.csv")
test_df  = pd.read_csv("hf://datasets/DGurgurov/slovak_sa/test.csv")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [5]:
#ensure values have correct types
for df in (train_df, val_df, test_df):
    df["text"] = df["text"].astype(str)
    df["label"] = df["label"].astype(int)

Lematizácia SlovakSA pre extrakciu features

In [6]:
!pip -q install stanza

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 61.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 773.7/773.7 kB 32.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 608.4/608.4 kB 44.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 337.2/337.2 kB 21.3 MB/s eta 0:00:00


In [7]:
import stanza
stanza.download("sk", verbose=False)

#initialize stanza tokenizer
nlp = stanza.Pipeline(
    lang="sk",
    processors="tokenize,pos,lemma",
    tokenize_no_ssplit=True,
    use_gpu=True,
    verbose=False
)

In [9]:
def lemmatize_text(df, text_col="text", out_col="lemma_text"):
    df_lemma = df.copy(deep=True)

    lemma_texts = []

    for text in df_lemma[text_col].fillna("").astype(str):
        doc = nlp(text)

        lemmas = []
        for sent in doc.sentences:
            for w in sent.words:
                if w.lemma:
                    lemmas.append(w.lemma.lower().replace(" ", "_"))

        lemma_texts.append(" ".join(lemmas))

    df_lemma[out_col] = lemma_texts
    return df_lemma

In [10]:
train_lem = lemmatize_text(train_df)
val_lem   = lemmatize_text(val_df)
test_lem  = lemmatize_text(test_df)

train_lem[[ "text", "lemma_text" ]].head()

,text,lemma_text
0,"Bola ústretová,milá a komunikatívna .Šikovná s...","byť ústretový , milý a komunikatívny . šikovný..."
1,"dobrá obsluha, pekná baba.","dobrý obsluha , pekný baba ."
2,"Prijemne vystupovanie, ochota pomôcť","prijemný vystupovanie , ochota pomôcť"
3,Za ochotu,za ochot
4,Ďakujem za rýchle vybavenie a ústretovosť,ďakovať za rýchly vybavenie a ústretovosť


#Základný RoBERTa model (Fine-Tuning)

Konverzia Dataframe na DatasetDict

In [11]:
from datasets import Dataset, DatasetDict

dset = DatasetDict({
    "train": Dataset.from_pandas(train_df, preserve_index=False),
    "validation": Dataset.from_pandas(val_df, preserve_index=False),
    "test": Dataset.from_pandas(test_df, preserve_index=False),
})

dset

DatasetDict({
    train: Dataset({
        features: ['label', 'text'],
        num_rows: 3560
    })
    validation: Dataset({
        features: ['label', 'text'],
        num_rows: 522
    })
    test: Dataset({
        features: ['label', 'text'],
        num_rows: 1042
    })
})

Konvertovali sme SlovakSA pandas Dataframy na DatasetDict, aby boli dáta vo vhodnom formáte na trénovanie modelu.

Konverzia textu na input_ids a attention_mask pre RoBERTa model

In [13]:
from transformers import AutoTokenizer

MODEL_NAME = "TUKE-KEMT/slovak-roberta-base"
#use RoBERTa tokenizer to ensure compatibility
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)

def tokenize(batch):
    return tokenizer(
        batch["text"],
        truncation=True,
        max_length=256
    )

tok = dset.map(tokenize, batched=True)

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/772 [00:00<?, ?B/s]

Map:   0%|          | 0/3560 [00:00<?, ? examples/s]

Map:   0%|          | 0/522 [00:00<?, ? examples/s]

Map:   0%|          | 0/1042 [00:00<?, ? examples/s]

In [14]:
tok["train"][0].keys()

dict_keys(['label', 'text', 'input_ids', 'attention_mask'])

Vidíme, že máme atribúty input_ids a attention_mask, extrahované z textu SlovakSA, ktoré priamo použijeme na tréning modelu.

In [15]:
#prepare for model training
import numpy as np
import evaluate

from transformers import (
    AutoModelForSequenceClassification,
    TrainingArguments,
    Trainer,
    DataCollatorWithPadding
)

In [16]:
MODEL_NAME = "TUKE-KEMT/slovak-roberta-base"

#import model
model = AutoModelForSequenceClassification.from_pretrained(
    MODEL_NAME,
    num_labels=2 #binary classification
)

#prepare collator for input padding
collator = DataCollatorWithPadding(tokenizer)

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaForSequenceClassification LOAD REPORT from: TUKE-KEMT/slovak-roberta-base
Key                        | Status     | 
---------------------------+------------+-
lm_head.bias               | UNEXPECTED | 
lm_head.dense.bias         | UNEXPECTED | 
lm_head.layer_norm.bias    | UNEXPECTED | 
lm_head.layer_norm.weight  | UNEXPECTED | 
lm_head.dense.weight       | UNEXPECTED | 
classifier.out_proj.bias   | MISSING    | 
classifier.dense.weight    | MISSING    | 
classifier.out_proj.weight | MISSING    | 
classifier.dense.bias      | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Vygenerovali sme input_ids a attention_mask, avšak majú rôzne dĺžky. Preto použijeme collator, ktorý zabezpečí, že všetky vstupy majú identickú dĺžku (256)

Metriky

In [17]:
acc = evaluate.load("accuracy")
f1  = evaluate.load("f1")

#function to calculate metrics for model evaluation
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    preds = np.argmax(logits, axis=-1)
    return {
        "accuracy": acc.compute(predictions=preds, references=labels)["accuracy"],
        "f1": f1.compute(predictions=preds, references=labels, average="binary")["f1"],
    }

Konfigurácia trénovania

In [18]:
from transformers import TrainingArguments

#configure training parameters
args = TrainingArguments(
    output_dir="roberta_only",      #save directory
    learning_rate=2e-5,
    per_device_train_batch_size=16, #train on batches of 16
    per_device_eval_batch_size=32,  #evaluate batches of 32
    num_train_epochs=3,             #3 training cycles
    weight_decay=0.01,

    eval_strategy="epoch",          #evaluate after each epoch
    save_strategy="epoch",          #save after each epoch
    load_best_model_at_end=True,
    metric_for_best_model="f1",     #choose best model based on F1 score

    logging_steps=50,
    fp16=True,
)

In [19]:
from transformers import Trainer

#prepare trainer based on previously defined parameters
trainer = Trainer(
    model=model,
    args=args,
    train_dataset=tok["train"],
    eval_dataset=tok["validation"],
    data_collator=collator,
    compute_metrics=compute_metrics,
    processing_class=tokenizer,
)

In [20]:
#begin training
trainer.train()

Epoch,Training Loss,Validation Loss,Accuracy,F1
1,0.072545,0.103003,0.969349,0.982533
2,0.072651,0.094201,0.977011,0.986667
3,0.008725,0.114837,0.975096,0.985507


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

There were missing keys in the checkpoint model loaded: ['roberta.embeddings.LayerNorm.weight', 'roberta.embeddings.LayerNorm.bias', 'roberta.encoder.layer.0.attention.output.LayerNorm.weight', 'roberta.encoder.layer.0.attention.output.LayerNorm.bias', 'roberta.encoder.layer.0.output.LayerNorm.weight', 'roberta.encoder.layer.0.output.LayerNorm.bias', 'roberta.encoder.layer.1.attention.output.LayerNorm.weight', 'roberta.encoder.layer.1.attention.output.LayerNorm.bias', 'roberta.encoder.layer.1.output.LayerNorm.weight', 'roberta.encoder.layer.1.output.LayerNorm.bias', 'roberta.encoder.layer.2.attention.output.LayerNorm.weight', 'roberta.encoder.layer.2.attention.output.LayerNorm.bias', 'roberta.encoder.layer.2.output.LayerNorm.weight', 'roberta.encoder.layer.2.output.LayerNorm.bias', 'roberta.encoder.layer.3.attention.output.LayerNorm.weight', 'roberta.encoder.layer.3.attention.output.LayerNorm.bias', 'roberta.encoder.layer.3.output.LayerNorm.weight', 'roberta.encoder.layer.3.output.Laye

TrainOutput(global_step=669, training_loss=0.06372950367507023, metrics={'train_runtime': 149.3462, 'train_samples_per_second': 71.512, 'train_steps_per_second': 4.48, 'total_flos': 523804777898880.0, 'train_loss': 0.06372950367507023, 'epoch': 3.0})

In [41]:
#evaluate trained model on test dataset
test_metrics = trainer.predict(tok["test"])
test_metrics.metrics

{'test_loss': 0.1030028760433197,
 'test_accuracy': 0.9798464491362764,
 'test_f1': 0.9884551951621771,
 'test_runtime': 2.1997,
 'test_samples_per_second': 473.708,
 'test_steps_per_second': 15.002}

Na výsledku vidíme, že f1 score je dosť vysoké (0.986).

Pri trénovaní sme použili základný RoBERTa model, pričom sme nepoužili lex_feats. Model sa naučil binárnu klasifikáciu sentimentu len na základe textu. Ako ďaľší krok rozšírime model o lexikálne features a zistíme, či je model presnejší.

#Hybridný model RoBERTa. Základný model + lexicon features.

SentiWordNet SK

In [22]:
file_path = "SentiWordNet_3.0.0_SK.txt"
rows = []

#parse each row
with open(file_path, "r", encoding="utf-8") as f:
    for line in f:
        if line.startswith("#"):
            continue
        parts = line.strip().split("\t")
        if len(parts) != 6:
            continue
        #assign values from line
        pos, sid, pos_score, neg_score, synset_terms, gloss = parts

        #add values to list
        rows.append({
            "pos": pos,
            "sid": str(sid).zfill(8),
            "pos_score": float(pos_score),
            "neg_score": float(neg_score),
            "obj_score": 1.0 - (float(pos_score) + float(neg_score)),
            "synset_terms": synset_terms,
            "gloss": gloss
        })

#create dataframe
swn_syn = pd.DataFrame(rows)
swn_syn["synset_id"] = swn_syn["pos"] + ":" + swn_syn["sid"]
swn_syn = swn_syn.drop_duplicates("synset_id").set_index("synset_id")

In [23]:
swn_syn.head()

,pos,sid,pos_score,neg_score,obj_score,synset_terms,gloss
synset_id,,,,,,,
a:00001740,a,00001740,0.125,0.00,0.875,schopné#1,"""(usually followed by `to') having the necessa..."
a:00002098,a,00002098,0.000,0.75,0.250,neschopný#1,"""(usually followed by `to') not having the nec..."
a:00002312,a,00002312,0.000,0.00,1.000,dorzálne#2 abaxiálne#1,"""facing away from the axis of an organ or orga..."
a:00002527,a,00002527,0.000,0.00,1.000,ventrálna#2 adaxiálne#1,"""nearest to or facing toward the axis of an or..."
a:00002730,a,00002730,0.000,0.00,1.000,akroskopické#1,facing or on the side toward the apex


Exportujeme DataFrame slovenkého SWN pre ďaľšie použitie

In [24]:
swn_syn.to_csv("SlovakSWN_DF.csv", encoding="utf-8", index=True)

Príprava lexikónu pre modelovanie (výpočet polarity, explode, odstránenie lablov, agregácia synsetov)

In [25]:
#calculate polarity for each synset
swn_syn["polarity"] = swn_syn["pos_score"] - swn_syn["neg_score"]

#separate synset terms
tmp = swn_syn[["synset_terms", "polarity"]].copy()
#one synset term per row
tmp["term"] = tmp["synset_terms"].str.split()
tmp = tmp.explode("term")

#remove sense numbers
tmp["word"] = tmp["term"].str.split("#").str[0].str.lower().str.strip()
tmp = tmp.dropna(subset=["word"])
tmp = tmp[tmp["word"] != ""]

#aggregate synsets
lex_agg = tmp.groupby("word", as_index=False)["polarity"].mean()
lex_dict = dict(zip(lex_agg["word"], lex_agg["polarity"]))

print("Synsets:", len(swn_syn))
print("Unique words in lexicon:", len(lex_dict))

Synsets: 117659
Unique words in lexicon: 134847


Výpočet features pre vstupný dataset SlovakSA na základe lexikónu.

In [28]:
import re
#regex to take only alphabet characters
WORD_RE = re.compile(r"[A-Za-zÁÄČĎÉÍĹĽŇÓÔŔŠŤÚÝŽáäčďéíĺľňóôŕšťúýž]+", re.UNICODE)

#if text is string, find all matching regex
def lex_tokens(text: str):
    return WORD_RE.findall(text.lower()) if isinstance(text, str) else []

#extract 10 features from text
def lex_features(text: str, strong_thr: float = 0.5):
    toks = lex_tokens(text)
    n = len(toks)

    #safety check
    if n == 0:
        return np.zeros(10, dtype=np.float32)

    #take values from lexicon for each match
    vals = [lex_dict[t] for t in toks if t in lex_dict]

    #if no match was found, fill with zeroes
    if not vals:
        return np.zeros(10, dtype=np.float32)

    #assign positive and negative values
    vals = np.array(vals, dtype=np.float32)
    pos = vals[vals > 0]
    neg = vals[vals < 0]

    #calculate features
    return np.array([
        vals.sum(),                     #polarity sum
        vals.mean(),                    #polarity mean
        len(vals),                      #hits
        len(vals) / n,                  #coverage
        pos.sum() if pos.size else 0,   #sum positive
        neg.sum() if neg.size else 0,   #sum negative
        pos.max() if pos.size else 0,   #max positive
        neg.min() if neg.size else 0,   #max negative
        (pos >= strong_thr).sum(),      #max strong positive
        (neg <= -strong_thr).sum()      #max strong negative
    ], dtype=np.float32)


#apply to lemmatized SlovakSA dataframes
for df in (train_lem, val_lem, test_lem):
    df["lex_feats"] = df["lemma_text"].apply(lambda t: lex_features(t).tolist())

train_lem.head(10)

,label,text,lemma_text,lex_feats
0,1,"Bola ústretová,milá a komunikatívna .Šikovná s...","byť ústretový , milý a komunikatívny . šikovný...","[0.6289682388305664, 0.08985260874032974, 7.0,..."
1,1,"dobrá obsluha, pekná baba.","dobrý obsluha , pekný baba .","[1.0018938779830933, 0.2504734694957733, 4.0, ..."
2,1,"Prijemne vystupovanie, ochota pomôcť","prijemný vystupovanie , ochota pomôcť","[0.1845238208770752, 0.0615079402923584, 3.0, ..."
3,1,Za ochotu,za ochot,"[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ..."
4,1,Ďakujem za rýchle vybavenie a ústretovosť,ďakovať za rýchly vybavenie a ústretovosť,"[0.011408727616071701, 0.0028521819040179253, ..."
5,1,Ochotný personál.,ochotný personál .,"[0.34375, 0.171875, 2.0, 1.0, 0.34375, 0.0, 0...."
6,1,"Je vždy milá a ochotná poradiť a pomôcť, ja c...","byť vždy milý a ochotný poradiť a pomôcť , ja ...","[0.9598214626312256, 0.0479910746216774, 20.0,..."
7,1,príjemný personál,príjemný personál,"[0.2678571343421936, 0.1339285671710968, 2.0, ..."
8,1,Vyšiel mi v ústrety a hoci moja požiadavka neb...,vyjsť ja v ústrety a hoci môj požiadavka byť v...,"[-0.32003965973854065, -0.02133597806096077, 1..."
9,1,vzdy volny stojan a prijemny personal,vzdy volny stojan a prijemny personal,"[-0.08472222089767456, -0.04236111044883728, 2..."


Predchádzajúcou časťou kódu sme extrahovali "features" z textov SlovakSA. Pre každý text, ktorý mal zhodu v lexikóne sme vypočítali vektor lex_feats, ktorý obsahuje 10 desatinných čísiel.

In [29]:
train_df["lex_feats"] = train_lem["lex_feats"].values
val_df["lex_feats"]   = val_lem["lex_feats"].values
test_df["lex_feats"]  = test_lem["lex_feats"].values

Výpočet features sme robili na lematizovanom datasete, aby sme dosiahli viacej "hits" lexikónom. Následne sme skopírovali lex_feats do pôvodných datasetov, ktoré použijeme v hybridnom modeli.

In [30]:
train_df.to_csv("slovakSA_lemma_train_lex.csv", index=False)
val_df.to_csv("slovakSA_lemma_val_lex.csv", index=False)
test_df.to_csv("slovakSA_lemma_test_lex.csv", index=False)

Exportovali sme pôvodné SlovakSA s vypočítanými lex_feats pre ďaľšie použitie

Konverzia na DatasetDict

In [31]:
from datasets import Dataset, DatasetDict

dset_hybrid = DatasetDict({
    "train": Dataset.from_pandas(train_df, preserve_index=False),
    "validation": Dataset.from_pandas(val_df, preserve_index=False),
    "test": Dataset.from_pandas(test_df, preserve_index=False),
})

dset_hybrid

DatasetDict({
    train: Dataset({
        features: ['label', 'text', 'lex_feats'],
        num_rows: 3560
    })
    validation: Dataset({
        features: ['label', 'text', 'lex_feats'],
        num_rows: 522
    })
    test: Dataset({
        features: ['label', 'text', 'lex_feats'],
        num_rows: 1042
    })
})

In [42]:
tok = dset_hybrid.map(tokenize, batched=True)

Map:   0%|          | 0/3560 [00:00<?, ? examples/s]

Map:   0%|          | 0/522 [00:00<?, ? examples/s]

Map:   0%|          | 0/1042 [00:00<?, ? examples/s]

In [33]:
#hybrid collator for converting lex_feats to tensors
class HybridCollator:
    def __init__(self, tokenizer):
        self.base = DataCollatorWithPadding(tokenizer)

    def __call__(self, features):
        lex = [f["lex_feats"] for f in features]

        #remove lex_feats before padding
        features2 = []
        for f in features:
            g = dict(f)
            g.pop("lex_feats", None)
            features2.append(g)

        batch = self.base(features2)
        batch["lex_feats"] = torch.tensor(lex, dtype=torch.float32)
        return batch

hybrid_collator = HybridCollator(tokenizer)

In [34]:
import torch
import torch.nn as nn
from transformers import AutoModel

In [35]:
#hybrid RoBERTa model
class RobertaHybrid(nn.Module):
    def __init__(self, model_name, lex_dim=10, num_labels=2):
        super().__init__()
        self.encoder = AutoModel.from_pretrained(model_name) #load TUKE-KEMT/slovak-roberta-base
        hidden = self.encoder.config.hidden_size             #define number of hidden layers, 768

        self.classifier = nn.Linear(hidden + lex_dim, num_labels) #classifier has 768 + 10 hidden layers, and 2 for binary classification

    def forward(self, input_ids=None, attention_mask=None, lex_feats=None, labels=None):
        outputs = self.encoder(input_ids=input_ids, attention_mask=attention_mask)
        cls = outputs.last_hidden_state[:, 0, :]  # CLS token

        x = torch.cat([cls, lex_feats], dim=1) #merge with lex_feats
        logits = self.classifier(x) #calculate logits

        #calculate training loss
        loss = None
        if labels is not None:
            loss = nn.functional.cross_entropy(logits, labels)

        return {"loss": loss, "logits": logits}

In [36]:
hybrid_model = RobertaHybrid(MODEL_NAME, lex_dim=10, num_labels=2)

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

RobertaModel LOAD REPORT from: TUKE-KEMT/slovak-roberta-base
Key                       | Status     | 
--------------------------+------------+-
lm_head.bias              | UNEXPECTED | 
lm_head.dense.bias        | UNEXPECTED | 
lm_head.layer_norm.bias   | UNEXPECTED | 
lm_head.layer_norm.weight | UNEXPECTED | 
lm_head.dense.weight      | UNEXPECTED | 
pooler.dense.bias         | MISSING    | 
pooler.dense.weight       | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


In [37]:
#define hybrid trainer parameters
hybrid_trainer = Trainer(
    model=hybrid_model,
    args=args,                      #reuse arguments
    train_dataset=tok["train"],
    eval_dataset=tok["validation"],
    data_collator=hybrid_collator,
    compute_metrics=compute_metrics,
    processing_class=tokenizer,
)

In [38]:
hybrid_trainer.train()

Epoch,Training Loss,Validation Loss,Accuracy,F1
1,0.059807,0.157011,0.965517,0.980519
2,0.049671,0.107341,0.971264,0.983240
3,0.000828,0.111609,0.973180,0.984340


TrainOutput(global_step=669, training_loss=0.057524956829612034, metrics={'train_runtime': 136.953, 'train_samples_per_second': 77.983, 'train_steps_per_second': 4.885, 'total_flos': 0.0, 'train_loss': 0.057524956829612034, 'epoch': 3.0})

In [39]:
#evaluate trained model on test dataset
hybrid_test_metrics = hybrid_trainer.predict(tok["test"])
hybrid_test_metrics.metrics

{'test_loss': 0.1141318753361702,
 'test_accuracy': 0.9779270633397313,
 'test_f1': 0.9873556899395272,
 'test_runtime': 2.0941,
 'test_samples_per_second': 497.592,
 'test_steps_per_second': 15.759}

Hybridný model neprináša pozorovateľné rozdiely od základného modelu. To znamená, že základny model už extrahuje všetky textové features. Preto lexikálne rozšírenie neprinieslo novú informáciu do modelu. Z výsledkov môžeme usúdiť, že lexikálne features nezlepšili výkon modelu RoBERTa.